API/ev3_timer.ts.c

```diff
@@ -0,0 +1,208 @@
+ * \note split into a separate file, added better error handling & locking
+ */
+
+#include "ev3_timer.h"
+#include "ev3_timer.private.h"
+
+////////////////////////////////////////////////////////////////
+// IMPLEMENTATION
+
+// TIMER SLOTS //
+
+unsigned long TimerWait(unsigned long Time)
+{
+    return TimerGetMS() + Time;
+}
+
+void TimerReady(unsigned long Timer)
+{
+    // wait until timer has elapsed
+    // do not block other threads
+    while (Timer > TimerGetMS())
+    {
+        usleep(1000);
+    }
+}
+
+unsigned long Timer(uint8_t Timer)
+{
+    // slow timer 10 ticks per second
+    return FastTimer(Timer) / 10;
+}
+
+unsigned long FastTimer(uint8_t Timer)
+{
+    if (!TimerInitialized())
+        return 0UL;
+
+    // centisecond timer (100 ticks per second)
+    if (Timer < NUM_CS_TIMERS)
+    {
+        LOCK_OR_RETURN(timestamp, 0UL);
+        unsigned long currentValue = tcb.csTimers[Timer];
+        UNLOCK(timestamp);
+
+        return TimerGetCS() - currentValue;
+    }
+    return 0;
+}
+
+void ClearTimer(uint8_t Timer)
+{
+    return SetTimer(Timer, 0UL);
+}
+
+void SetTimer(uint8_t Timer, unsigned long Value)
+{
+    if (!TimerInitialized())
+        return;
+
+    if (Timer < NUM_CS_TIMERS)
+    {
+        LOCK_OR_RETURN(timestamp, /**/);
+        tcb.csTimers[Timer] = TimerGetCS() - Value;
+        UNLOCK(timestamp);
+    }
+}
+
+unsigned long TimerMS(uint8_t Timer)
+{
+    if (!TimerInitialized())
+        return 0UL;
+
+    if (Timer < NUM_MS_TIMERS)
+    {
+        LOCK_OR_RETURN(timestamp, 0UL);
+        unsigned long currentValue = tcb.msTimers[Timer];
+        UNLOCK(timestamp);
+
+        return TimerGetMS() - currentValue;
+    }
+    return 0;
+}
+
+void ClearTimerMS(uint8_t Timer)
+{
+    return SetTimerMS(Timer, 0UL);
+}
+
+void SetTimerMS(uint8_t Timer, unsigned long Value)
+{
+    if (!TimerInitialized())
+        return;
+
+    if (Timer < NUM_MS_TIMERS)
+    {
+        LOCK_OR_RETURN(timestamp, /**/);
+        tcb.msTimers[Timer] = TimerGetMS() - Value;
+        UNLOCK(timestamp);
+    }
+}
+
+unsigned long TimerUS(uint8_t Timer)
+{
+    if (!TimerInitialized())
+        return 0UL;
+
+    if (Timer < NUM_US_TIMERS)
+    {
+        LOCK_OR_RETURN(timestamp, 0UL);
+        unsigned long currentValue = tcb.usTimers[Timer];
+        UNLOCK(timestamp);
+
+        return TimerGetUS() - currentValue;
+    }
+    return 0;
+}
+
+void ClearTimerUS(uint8_t Timer)
+{
+    return SetTimerUS(Timer, 0UL);
+}
+
+void SetTimerUS(uint8_t Timer, unsigned long Value)
+{
+    if (!TimerInitialized())
+        return;
+
+    if (Timer < NUM_US_TIMERS)
+    {
+        LOCK_OR_RETURN(timestamp, /**/);
+        tcb.usTimers[Timer] = TimerGetUS() - Value;
+        UNLOCK(timestamp);
+    }
+}
+
+// CURRENT TIME QUERIES //
+
+unsigned long long TimerGetUS()
+{
+    struct timeval tv;
+    // this is documented not to fail
+    gettimeofday(&tv, 0);
+    return (((unsigned long long) tv.tv_sec) * 1000000) + tv.tv_usec;
```

Comment:
Just a nitpick: 1000000ULL saves you the parenthesis and thus makes code easier to read sometimes.

Reply:
I agree, I will change it.

Reply:
Resolved in 8099053b7a28ab2bd53675765d11dacd3aaa891d :+1: